OP-759: Restructuring package management

TomVasile opened the pull request:
### Overview
WIP

### Which JIRA ticket does this PR relate to?
Covers two:
https://casperlabs.atlassian.net/browse/OP-759
https://casperlabs.atlassian.net/browse/OP-806

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [ ] If this PR adds a new feature, it includes tests related to this feature.
- [ ] You assigned one person to review this PR.
- [ ] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.
- [ ] Do not forget to run `bors r+` if GitHub policy is not enforced, e.g. when merging into another feature branch. It may be omitted under some circumstances if this PR intentionally assumes that integration tests will fail but will be fixed with the future PRs.

### Notes
- Pre-install of node will create the `/etc/casperlabs` directory. This will allow the techspec to be move from inside the jar and unpack in the `/etc/casperlabs` directory.
- Post-install takes care of ensuring the default data directory is present: `/var/lib/casperlabs`

aakoshh commented on node/src/main/resources/default-configuration.toml:
```diff
@@ -31,7 +31,7 @@ default-timeout = "5second"
 
 # Path to data directory.
-data-dir = "$HOME/.casperlabs"
+data-dir = "/var/lib/casperlabs"
```
Isn't this going to break it for anyone trying to run it from the tarball?

TomVasile replied:
The default toml is broken no matter, I plan to add a document that explains how this file can be used.

aakoshh replied:
Step 5 and 6 here assume that the stuff is in home I think: https://github.com/CasperLabs/CasperLabs/blob/dev/docs/NODE.md#step-5-start-the-execution-engine

TomVasile replied:
those are overidding the default toml, if someone was to use the toml they need to specify `--config-file=`

aakoshh replied:
Everything is using the deault toml file. It's packed into the software and if you don't specify any values, this is where they are ultimately coming from.

TomVasile replied:
reverted

aakoshh commented on node/src/main/resources/default-configuration.toml:
```diff
@@ -158,18 +158,18 @@ use-tls = false
 
 # Private key to use with the public API certificate.
-api-key = "$HOME/.casperlabs/api.key.pem"
+#api-key = "$HOME/.casperlabs/api.key.pem"
```
Removing these default options will possibly mean some examples will start failing. Although it's possible they have already been made obsolete when I added the separate API keys.

aakoshh replied:
Just noticed https://github.com/CasperLabs/CasperLabs/blob/dev/docs/NODE.md#step-3-start-the-node has motes, and mint/pos wasm too, so definitely broken already.

TomVasile replied:
i also dont believe these options are documented anywhere. They aren't required to run a node so I commented them out.

aakoshh replied:
The files in `Tls.scala` are non-optional so these values were used, unless an override was provided. If you remove it and an explicit value is not provided the node will not start.

TomVasile replied:
then why can i start a node with out supplying these values and also not having api.key.pem in my data dir?

aakoshh replied:
As discussed on Slack, the reason it hasn't been a problem for you is because you had gRPC encryption of the public API turned off, in which case it won't try to use these files: https://github.com/CasperLabs/CasperLabs/blob/dev/node/src/main/scala/io/casperlabs/node/NodeRuntime.scala#L103

TomVasile replied:
reverted

aakoshh commented on node/src/main/resources/default-configuration.toml:
```diff
@@ -204,7 +204,7 @@ validator-sig-algorithm = "ed25519"
 
 # Path to the directory which contains the Chain Spec. If not specified, the default chainspec packaged with the node will be used.
-# chain-spec-path = "$HOME/.casperlabs/chainspec"
+# chain-spec-path = "/etc/casperlabs/chainspec"
```
Do you think we could leave the values as as $HOME based and just overwrite in the service spec?

aakoshh replied:
Ah but we can't if this is the file that gets copied over there. Never mind.

TomVasile replied:
yea we need to move away from using the home directory and start following the linux filesystem standard.

ie. `/etc/` for config files. `/var/lib` for data.  et-cetra

TomVasile commented:
bors try

arcolife commented on node/src/main/resources/default-configuration.toml:
```diff
@@ -142,7 +142,7 @@ cache-neighborhood-after = 6
 
 # Socket path used for internal gRPC API.
-socket = "$HOME/.casperlabs/.casper-node.sock"
+socket = "/var/lib/casperlabs/.casper-node.sock"
```
is this write-able by `casperlabs` user?

TomVasile replied:
yes, this is taken care of in the post-install maintainer scripts

arcolife replied:
right I see it now. 

So another one: If I stop the service, choose to run it as a different user, my data-dir/socket files are also affected? Thereby affecting boot state of the next time I bring it up?

arcolife replied:
or re-wording: If i choose to run  it as a different user, is there a reference script I can use to do this? Shouldn't we include that as well? 
or should we make the data/socket file paths user-agnostic? and then on restart, have a pre-script to `execstart` that changes the file / ownerships?

TomVasile replied:
correct, but theres no way around that. We can only give them a default way to run. If they want to modify that in anyway its up to them

arcolife replied:
but I don't think linking username to data dir is the correct way to go? maybe just the ownership, sure (and the admin would have to take care of that, but not the folder/socket names?)

TomVasile replied:
i dont think changing file ownership is standard practice nor something we should ever do.

TomVasile replied:
if the users want to run as a different user that is up to them. We just need to supply a default way to run and if powerusers want to change that they can feel free.

arcolife replied:
so scratch ownership, but what about data-dir? I guess the idea currently here is if we switch the user, the script would complain along the lines of "/var/lib/new_user" doesn't exist. And then the admin would go do that? If so, looks fine, but there's a chance of confusion here?

TomVasile replied:
i don't think there will be confusion once I provide the doc explaining the default configuration and how user will need to modify that if they want to run differently.

The goal here is to provide a working way by **default**. I don't believe what you are describing is default behavior.

arcolife replied:
right the doc should be fine.

> So another one: If I stop the service, choose to run it as a different user, my data-dir/socket files are also affected? Thereby affecting boot state of the node, next time I bring it up?

^ but repeating myself here, I believe the standard behavior is to have a hardcoded product name like `/var/lib/casperlabs` instead of adding a username under `/var/lib` for datadir.

TomVasile replied:
> ^ but repeating myself here, I believe the standard behavior is to have a hardcoded product name like /var/lib/casperlabs

it is, its hardcoded to `/var/lib/casperlabs`

> instead of adding a username under `/var/lib` for datadir

its not "adding a username under `/var/lib`", its adding the default directory which happens to be the same name as the user.

> ref: http://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/Linux-Filesystem-Hierarchy.html#var
Users shouldn't ever need to modify files in `/var/lib` to configure a package's operation

I don't see a reason for them to need to touch `/var/lib` either, hence this setup for them.

arcolife replied:
alright. My only concern was change of user / other user by default, but that should be covered by doc, so hey whatever works!

TomVasile commented:
> What I thought would also happen is changing the Dockerfiles to install the packages.
That is coming separately.

> I think some of the changes in default paths from $HOME to /etc means that the hack/docker setup will most likely not work any more. Docs need updating too. Note sure about integration tests.

I don't think these changes will break anything as nobody uses the toml file. I don't see it being used by the hack/docker setup. Can you point me to where?

The changes to the default-toml is meant to serve as an example and be overwritten by admins. I can add the doc to this PR to help explain my thought process here but a quick tldr:

This file can never fully be packaged in a working state because there are parameters that can't be set by default and will need to be changed. ie. keys, bootstrap address, etc.

TomVasile commented:
bors try

TomVasile commented:
bors try

aakoshh commented:
Everything is depending on the settings in `default-configuration.toml` because of https://github.com/CasperLabs/CasperLabs/blob/dev/node/src/main/scala/io/casperlabs/node/configuration/Configuration.scala#L119

aakoshh commented:
I don't want to have hack/docker left broken. If it has to be done separately, then these changes should go on a feature branch.

TomVasile commented:
> Everything is depending on the settings in `default-configuration.toml` because of https://github.com/CasperLabs/CasperLabs/blob/dev/node/src/main/scala/io/casperlabs/node/configuration/Configuration.scala#L119

if this is the case then I'm happy to fix hack/docker, int tests etc.

TomVasile commented:
spoke with akosh on slack, I'll need to tie down the loose ends on this PR now that I have a better understanding of the role the toml plays. Reviewers can hold-off on this for a bit.

TomVasile commented:
@aakoshh I think this guy is ready now. I reverted all of the edits to the default toml and switched it to package an example of the toml which users can edit to use. Included a doc as well for users not super familiar with systemd but still want to use it this way.

aakoshh commented on docs/SYSTEMD.md:
```diff
@@ -0,0 +1,70 @@
+bootstrap address, etc.), we've opted to give users an *almost* production 
+ready version of these systemd unit files. The node unit file references an 
+`example-configuration.toml` which should be copied and edited to reflect paths
+ paths for end-user specific changes. The `example-configuration.toml` lives 
```
```suggestion
 for end-user specific changes. The `example-configuration.toml` lives 
```

TomVasile replied:
ty

aakoshh commented on docs/SYSTEMD.md:
```diff
@@ -0,0 +1,70 @@
+the CasperLabs software relies on end-user specific flags (for keypaths, 
+bootstrap address, etc.), we've opted to give users an *almost* production 
+ready version of these systemd unit files. The node unit file references an 
+`example-configuration.toml` which should be copied and edited to reflect paths
```
It's not clear to me where this should be copied. Should it be renamed?

TomVasile replied:
its described further below: https://github.com/CasperLabs/CasperLabs/pull/1291/files#diff-36c4552ce30bc51ca4719f00d23b0c37R25

aakoshh replied:
Yeah, the examples made it clear.

aakoshh commented on docs/SYSTEMD.md:
```diff
@@ -0,0 +1,70 @@
+
+These unit-files by default are stored under `/lib/systemd/system` and we recommend 
+that they are copied over to `/etc/systemd/system` so adequate changes can be 
+made for use.
```
I don't understand this part. Is this different from the `example-configuration.toml`?

aakoshh replied:
Maybe this could be written first, and then the example-config.

TomVasile replied:
These files are different from the `example-conf.toml` yes, as these are the systemd unit files. I'm unsure what moving this section accomplishes? 

They are described above in https://github.com/CasperLabs/CasperLabs/pull/1291/files#diff-36c4552ce30bc51ca4719f00d23b0c37R8-R11

and this statement is describing what I'm showing below in: https://github.com/CasperLabs/CasperLabs/pull/1291/files#diff-36c4552ce30bc51ca4719f00d23b0c37R23-R24

aakoshh replied:
In retrospect, after seeing the code below, I understand. I just thought it would be more natural to do it like this:

> With the `deb` and `rpm` packages of the node and execution engine we install 
default versions of systemd unit files that users can chose to utilize. Since 
the CasperLabs software relies on end-user specific flags (for keypaths, 
bootstrap address, etc.), we've opted to give users an *almost* production 
ready version of these systemd unit files.
>
> These unit-files by default are stored under `/lib/systemd/system` and we recommend 
that they are copied over to `/etc/systemd/system` so adequate changes can be 
made for use.
> 
> The node unit file references an `example-configuration.toml` which should be copied and edited to reflect paths for end-user specific changes. The `example-configuration.toml` lives under `/etc/casperlabs/` by default.

That way you tell first what the unit files are and where they are, and later the more specific configuration in one of them.

TomVasile replied:
added in https://github.com/CasperLabs/CasperLabs/pull/1291/commits/6a710dc5a4068ba3ddd945554d2b01b5e0eac7ab

aakoshh commented on docs/SYSTEMD.md:
```diff
@@ -0,0 +1,70 @@
+
+By default this file points to the `example-configuration.toml`. You will want
+want to update this in `/etc/systemd/system/casperlabs-node.service` to point 
+to the configuration you updated for youself. In the below, case I changed it 
```
```suggestion
to the configuration you updated for youself. In the below case I changed it 
```

TomVasile replied:
ty

aakoshh commented on build.sbt:
```diff
@@ -307,9 +307,11 @@ lazy val node = (project in file("node"))
     linuxPackageMappings ++= {
-      val file = baseDirectory.value / "casperlabs-node.service"
+      val service_file = baseDirectory.value / "casperlabs-node.service"
+      val example_toml = baseDirectory.value / "../packaging/example-configuration.toml"
```
I think this should either copy copy the `default-configuration.toml`, or the example should have all the values commented out so as not to interfere with default settings.

TomVasile replied:
so the reason I made a copy of this file is because I want the packaged version to follow FHS. ie. data-dir = "/var/lib/casperlabs" and not a users home directory

aakoshh replied:
I think you misunderstood. I didn't mean not to copy an example to wherever the FHS tradition dictates. I meant that we shouldn't have a "../packaging/example-configuration.toml" file, which is, today, a copy of `baseDirectory.value / "src/main/resources/default-configuration.toml"`, but tomorrow it will have obsolete values and fall out of sync. We should just copy the default itself.

Or, which I think would ultimately be better, copy a version which has all the settings commented out. Why? Because the user will make a `cp` once, and then modify 1 value, but in fact they'll have fixed all the other settings as well, even the ones they did not care about, and that will negate our ability to change those defaults in subsequent releases.

TomVasile replied:
added this in https://github.com/CasperLabs/CasperLabs/pull/1291/commits/ccea27ab4384593955c668cb80caf19dfd3aa047

quick explanation: we copy the default toml and then in the deb/rpm postinstall we comment out the whole file and switch the `$HOME` references to either `/var/lib/casperlabs` or `/etc/casperlabs`. This way, if users take the systemd path of running, they can just uncomment and run.

TomVasile commented:
> Looks good but the example will fall out of sync with the defaults. We should copy the default IMO, or comment out the value from the example. Commenting them out would also be helpful if the user copied it and modified one particular value, so that the other values they don't care about can still be changed in the defaults.

I went back an forth but ultimately I agree. I think I'll take the approach of copying the default and then commenting everything out in the postinstall scripts.

aakoshh commented:
> I went back an forth but ultimately I agree. I think I'll take the approach of copying the default and then commenting everything out in the postinstall scripts.

Ja Mann, that's the most advanced option!

piotr-putit commented on docs/SYSTEMD.md:
```diff
@@ -0,0 +1,74 @@
+#### CasperLabs Node Unit File
+
+By default this file points to the `example-configuration.toml`. You will want
+want to update this in `/etc/systemd/system/casperlabs-node.service` to point 
```
double want

TomVasile replied:
ty

arcolife commented on node/src/rpm/scriptlets/post:
```diff
@@ -2,22 +2,73 @@
+# Some sed magic on DEFAULT_TOML
+# Neat trick on SED_CHECK :)
+SED_CHECK=$(command -v sed >/dev/null 2>&1)$?
+if [ "$SED_CHECK" -eq "0"  ] && [ -f ${DEFAULT_TOML} ]; then
```
if `sed`'s not present, should the setup continue without defaults?

arcolife replied:
(anyway is sed ever not present on rpm based systems?)

TomVasile replied:
i've never come across a distro that didn't have sed. It's tough to support every distro out there and from discussions we've had in the past we only plan to officially support a few (ubuntu, fedora, etc). 

in this case, if sed is not available the defaults would remain the same since I'm copying the default toml. Would be a no-op for the user as these defaults are set in the packaging as well.

TomVasile commented:
bors try

TomVasile commented:
bors try

TomVasile commented:
bors try-

TomVasile commented:
bors try

TomVasile commented:
bors try-

TomVasile commented:
bors try

TomVasile commented:
bors r+
